hazelcast/src/main/java/com/hazelcast/map/impl/record/ObjectRecordFactory.java

```diff
@@ -51,17 +51,11 @@ public void setValue(Record<Object> record, Object value) {
 
     @Override
     public boolean isEquals(Object value1, Object value2) {
-        Object v1 = value1 instanceof Data ? serializationService.toObject(value1) : value1;
-        Object v2 = value2 instanceof Data ? serializationService.toObject(value2) : value2;
-        if (v1 == null && v2 == null) {
+        if (value1 == value2) {
             return true;
         }
-        if (v1 == null) {
-            return false;
-        }
-        if (v2 == null) {
-            return false;
-        }
-        return v1.equals(v2);
+        Object v1 = value1 instanceof Data ? serializationService.toObject(value1) : value1;
```

Comment:
Maybe as an optimisation, we could add the 
```
if (value1 == null || value2 == null) {
              return false;
          }
```
before deserialising to skip deserialising if one parameter is `null` but the other one isn't.

Reply:
Good one! I also added some more test cases with an empty `Data`, so the last comparison still has full coverage (the `toData()` can still return `null` with an empty `Data` instance).